torch/csrc/autograd/cpp_hook.h

```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include <torch/csrc/autograd/function_hook.h>
+#include <functional>
+#include <memory>
+
+namespace torch { namespace autograd {
+
+using hooks_list = std::vector<std::function<Variable(const Variable&)>>;
+
+struct CppFunctionPreHook : public FunctionPreHook {
+  CppFunctionPreHook(const std::shared_ptr<hooks_list> &hooks, int value_idx);
+  variable_list operator()(const variable_list& values) override;
+
+  std::shared_ptr<hooks_list> hooks_;
```

Comment:
Is there any reason this has to be a shared pointer to a vector, as opposed to just a vector?

Reply:
The variable and all the CppFunctionHooks need to be able to access it. This is similar to what we do in python with the backward_hooks dict.